compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2537,4 +2563,196 @@ public int hashCode() {
+            }
+        }
+        return true;
+    }
+
+    /**
+     * This will handle two types.
+     *  Singleton : As singleton can have one value that value should it self be a valid fill value
+     *  Union :
+     *          1. if nil is a member it is the fill values
+     *          2. else all the values should belong to same type and the default value for that type
+     *              should be a member of the union
+     * @param type BFiniteType union or finite
+     * @return
+     */
+    private boolean checkFillerValue(BFiniteType type) {
+        // For singleton types, that value is the implicit initial value
+        if (type.valueSpace.size() == 1) {
+            return true;
+        }
+
+        // is first value null
+        Iterator iterator = type.valueSpace.iterator();
+        BLangExpression firstElement = (BLangExpression) iterator.next();
+        if ((firstElement == null) || (firstElement.type.getKind() == TypeKind.NIL)) {
+            return true;
+        }
+
+        boolean allMembersHaveSameType = true;
+        boolean defaultFillValuePresent = false;
+
+        // is first value is a valid fill value
+        String defaultFillValue = getDefaultFillValue(firstElement);
+        if (firstElement.toString().equals(defaultFillValue)) {
+            defaultFillValuePresent = true;
+        }
+
+        while (iterator.hasNext()) {
+            Object value =  iterator.next();
+            if (value == null) {
+                return true;
+            }
+
+            BType valueType = ((BLangExpression) value).type;
+            if (valueType.getKind() == TypeKind.NIL) {
+                return true;
+            }
+
+            if (allMembersHaveSameType && !isSameType(valueType, firstElement.type)) {
+                allMembersHaveSameType = false;
+            }
+
+            if (!defaultFillValuePresent && value.toString().equals(defaultFillValue)) {
+                defaultFillValuePresent = true;
+            }
+        }
+
+        if (!allMembersHaveSameType) {
+            return false;
+        }
+
+        return defaultFillValuePresent;
+    }
+
+    private boolean checkFillerValue(BUnionType type) {
+        if (type.isNullable()) {
+            return true;
+        }
+
+        Iterator<BType> iterator = type.getMemberTypes().iterator();
+        BType firstMember = iterator.next();
+        // is first value is a valid fill value
+        if (firstMember.getKind() == TypeKind.NIL) {
+            return true;
+        }
+
+        boolean allMembersHaveSameType = true;
+        boolean defaultFillValuePresent = false;
+
+        // is first value is a valid fill value
+        String defaultFillValue = getDefaultFillValue(firstMember);
+        if (firstMember.toString().equals(defaultFillValue)) {
+            defaultFillValuePresent = true;
+        }
+
+        while (iterator.hasNext()) {
+
+            Object value =  iterator.next();
+            if (value == null) {
```

Comment:
Again, can these be null?

Reply:
Let me remove. I think this should not be null